keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,636 @@
+  // borrowing resources when this ClusterQueue has pending Workloads and the
+  // min quota is unmet.
+  preemption ClusterQueuePreemption
+}
+
+type ClusterQueuePreemption struct {
+  // waitBeforePreemptionSeconds is the time in seconds that the ClusterQueue
+  // is willing to wait for Workloads to finish before issuing preemptions,
+  // since the first time that the ClusterQueue has pending workloads that would
+  // fit under the min quota. This time is present in the timestamp for the
+  // PendingWorkloadsUnderMinQuota condition.
+  // If null, it means to wait indefinitely.
+  WaitBeforePreemptionSeconds *int64
+}
+```
+
+### Workload changes
+
+The new API field in Workload describe what to do in the case of a preemption.
+
+```golang
+type OnPreemptionAction string
+
+const (
+  OnPreemptionActionRequeue OnPreemptionAction = "Requeue"
+  OnPreemptionActionTerminate OnPreemptionAction = "Terminate"
+)
+
+type WorkloadSpec struct {
+  ...
+  // onPreemption describes what to do with the workload when it's preempted.
+  // Options are:
+  //
+  // - Requeue (default): The Workload is put back in the queue for admission.
+  //   Note that the original .metadata.CreationTimestamp would be used for
+  //   sorting this workload.
+  // - Terminate: The workload gets a Finished condition with reason Preempted.
+  OnPreemption OnPreemptionAction
+}
+```
+
+### Changes in scheduling algorithm
+
+The following changes in the scheduling algorithm are required to implement
+preemption.
+
+#### Detecting Workloads that might benefit from preemption
+
+The first stage during scheduling is to assign flavors to each resource of
+a workload.
+
+The algorithm is like follows:
+
+    For each resource (or set of resources with the same flavors), evaluate
+    flavors in the order established in the ClusterQueue:
+
+    0. Find a flavor that still has quota in the cohort (borrowing allowed),
+       but doesn't surpass the max quota for the CQ. Keep track of whether
+       borrowing was needed.
+    1. [New step] if no flavor was found, find a flavor that is under min quota
+       for the CQ, regardless of the cohort usage.
+    2. [New step] if no flavor was found, use the first flavor in the list that
+       has more quota than the Workload request.
+
+We will call the flavor assignments type 0, 1 or 2, depending on which step they
+were found.
+    
+A flavor assignment of type 1 means that we need to preempt or wait for other
+workloads in the cohort to finish to accomodate this workload, because the
+ClusterQueue is lending its resources.
+
+A flavor assigment of type 2 means that we need to preempt or wait for other
+workloads in the ClusterQueue to finish to accomodate this workload.
+
+Note that a Workload could get assignments of all types.
+
+#### Sorting among heads of ClusterQueues
+
+Sorting uses the following criteria:
+
+1. Flavors that don't borrow first.
+2. [New criteria] Highest priority first.
```

Comment:
if the high priority workload fits with borrowing, it can still borrow, but in the next scheduling cycle.

If there is no capacity to borrow, it will preempt low priority workloads, and they wouldn't fit when requeued either.

Reply:
> So no, there wouldn't be starvation within the ClusterQueue.

If we have several lower priority workloads which don't borrow resources and one hight priority workloads which needs to borrow resources. In my understanding, the hight priority workloads is starving, no?

Reply:
In scheduling, we prefer the workloads wouldn't borrow resources, but we support preemption via priority, aren't they conflict with each other, that means lower workloads doesn't borrow resources might be scheduled at first, but then they will be preempted for higher priority ones?

How about only care about the priority in preemption, at least for startup. And leave the flavor assignment to https://github.com/kubernetes-sigs/kueue/issues/312, like MinimizeBorrowing. I think this is more clear.

Reply:
This is not exactly the same as your design, but can somehow mitigate this.

Reply:
> If we have several lower priority workloads which don't borrow resources and one hight priority workloads which needs to borrow resources. In my understanding, the hight priority workloads is starving, no?

This is somewhat true. What you might be missing is that we are talking about heads of the ClusterQueues here.
So we have two heads of different ClusterQueues competing, but both have the highest priority within their ClusterQueue. Let's say A has lower priority than B. If B fits without borrowing, it should still be scheduled first, to satisfy the min quota of the ClusterQueue. That's the definition of min.

> we prefer the workloads wouldn't borrow resources, but we support preemption via priority, aren't they conflict with each other

No. Preemption happens in two scenarios: (1) within a cohort, to recover the min quota that was borrowed to other ClusterQueues or (2) within a ClusterQueue, purely due to priority.

> I think maybe in you design: workload_b will be scheduled firstly, right? If there are so many short-term workloads like workload_b, the workload_a will wait all the workloads like workload_b scheduled? right? Here may lead to some starvation that higher priority workload will wait for a long time?

Yes, this is working as intended. We should respect the min quota of the ClusterQueue for B first.
But workload_a could still preempt lower priority workloads within its ClusterQueue.